src/Stratis.Bitcoin.Features.PoA/Voting/IdleFederationMembersKicker.cs

```diff
@@ -0,0 +1,144 @@
+            FederationManager federationManager, SlotsManager slotsManager, VotingManager votingManager, ILoggerFactory loggerFactory)
+        {
+            this.signals = signals;
+            this.network = network;
+            this.keyValueRepository = keyValueRepository;
+            this.consensusManager = consensusManager;
+            this.federationManager = federationManager;
+            this.slotsManager = slotsManager;
+            this.votingManager = votingManager;
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.federationMemberMaxIdleTimeSeconds = ((PoAConsensusOptions)network.Consensus.Options).FederationMemberMaxIdleTimeSeconds;
+        }
+
+        public void Initialize()
+        {
+            this.blockConnectedToken = this.signals.Subscribe<BlockConnected>(this.OnBlockConnected);
+            this.fedMemberAddedToken = this.signals.Subscribe<FedMemberAdded>(this.OnFedMemberAdded);
+            this.fedMemberKickedToken = this.signals.Subscribe<FedMemberKicked>(this.OnFedMemberKicked);
+
+            this.fedMembersByLastActiveTime = this.keyValueRepository.LoadValueJson<Dictionary<PubKey, uint>>(fedMembersByLastActiveTimeKey);
+
+            if (this.fedMembersByLastActiveTime == null)
+            {
+                this.logger.LogDebug("No saved data found. Initializing federation data with genesis timestamps.");
+
+                this.fedMembersByLastActiveTime = new Dictionary<PubKey, uint>();
+
+                // Initialize federation members with genesis time.
+                foreach (PubKey federationMember in this.federationManager.GetFederationMembers())
+                    this.fedMembersByLastActiveTime.Add(federationMember, this.network.GenesisTime);
+
+                this.SaveMembersByLastActiveTime();
+            }
+        }
+
+        private void OnFedMemberKicked(FedMemberKicked fedMemberKickedData)
+        {
+            this.fedMembersByLastActiveTime.Remove(fedMemberKickedData.KickedMember);
+
+            this.SaveMembersByLastActiveTime();
+        }
+
+        private void OnFedMemberAdded(FedMemberAdded fedMemberAddedData)
+        {
+            if (!this.fedMembersByLastActiveTime.ContainsKey(fedMemberAddedData.AddedMember))
+            {
+                this.fedMembersByLastActiveTime.Add(fedMemberAddedData.AddedMember, this.consensusManager.Tip.Header.Time);
+
+                this.SaveMembersByLastActiveTime();
+            }
+        }
+
+        private void OnBlockConnected(BlockConnected blockConnectedData)
+        {
+            // Update last active time.
+            uint timestamp = blockConnectedData.ConnectedBlock.ChainedHeader.Header.Time;
+            PubKey key = this.slotsManager.GetPubKeyForTimestamp(timestamp);
+            this.fedMembersByLastActiveTime.AddOrReplace(key, timestamp);
+
+            this.SaveMembersByLastActiveTime();
```

Comment:
Save on every block?

Reply:
Yes. But it will be a fast save since there will be less than 100 items in the dictionary. 
But if you want I can optimize it by saving in a task which is ran on background and is awaited on next block before starting the save again.

Reply:
If you do that you'll have to employ the tips manager no?